File: drivers/python/examples/01-ingest-to-replay.py
"""examples/01-ingest-to-replay.py

Python mirror of examples/01-ingest-to-replay.sql. Run after
`pip install -e .` against a database that has maludb_core
installed:

    MALUDB_DSN="postgresql:///mydb" python examples/01-ingest-to-replay.py
"""

from __future__ import annotations

import os
import sys
import uuid

from maludb import MaluDBClient


def main() -> int:
    dsn = os.environ.get("MALUDB_DSN", "postgresql:///mydb")
    # Suffix the SVPOR signature with a per-run id. The
    # malu$fact_active_window_excl EXCLUDE constraint refuses two
    # active facts with the same (subject, verb, predicate) on
    # overlapping valid windows — that's the supersession engine
    # doctrine. Per-run subjects keep the example idempotent.
    run = uuid.uuid4().hex[:8]
    subject = f"api_gateway_{run}"

    with MaluDBClient.from_dsn(dsn) as client:
        print(f"connected to {dsn} — maludb_core {client.version()}")
        print(f"run-id = {run}, subject = {subject}")

        sp = client.register_source_package(
            source_type="log",
            content_text=f"py-example-01 [{run}]: 14:22Z api-gateway 5xx burst",
            origin={"uri": f"log://oncall/py-example-01/{run}"},
        )
        print(f"  source_package_id = {sp}")

        c1 = client.register_claim(
            subject=subject,
            verb="observed",
            object_value="5xx_burst",
            statement_text=f"py-example-01 [{run}]: initial 5xx surge at 14:22Z",
            source_package_id=sp,
        )
        c2 = client.register_claim(
            subject=subject,
            verb="timed_out",
            object_value="health_probe",
            statement_text=f"py-example-01 [{run}]: health probe exceeded 2s",
            source_package_id=sp,
        )
        print(f"  claim_ids = {c1}, {c2}")

        f1 = client.register_fact(
            claim_ids=[c1, c2],
            subject=subject,
            verb="incident",
            object_value="latency_breach",
            statement_text=f"py-example-01 [{run}]: latency SLO breach root cause",
            verification_method="oncall_review",
        )
        print(f"  fact_id = {f1}")

        ep = client.register_episode(
            episode_kind="incident",
            title=f"py-example-01-outage-{run}",
            summary="Driver example outage",
            payload={"subject_class": subject, "environment": "prod"},
        )
        print(f"  episode_id = {ep}")

        # Query
        print(f"\n=== text_search '{subject}' ===")
        for h in client.text_search(subject, limit=5):
            print(f"  {h.object_type:14}  id={h.object_id:5}  rank={h.rank:.4f}")

        print(f"\n=== retrieve '{subject}' ===")
        for h in client.retrieve(subject, limit=5):
            print(f"  {h.object_type:14}  id={h.object_id:5}  strategy={h.strategy}")

        print("\n=== replay_episode ===")
        envelope = client.replay_episode(ep, mode="current_valid")
        print(f"  mode={envelope['mode']}  step_count={len(envelope['steps'])}  "
              f"evidence={len(envelope['supporting_evidence'])}")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: drivers/nodejs/examples/01-ingest-to-replay.ts
/**
 * examples/01-ingest-to-replay.ts
 *
 * Node.js mirror of examples/01-ingest-to-replay.sql. Run via:
 *
 *   MALUDB_DSN=postgresql:///maludb_bench \
 *     node --import tsx examples/01-ingest-to-replay.ts
 */

import { randomBytes } from "node:crypto";
import { MaluDBClient } from "../src/index.js";

async function main(): Promise<void> {
  const dsn = process.env["MALUDB_DSN"] ?? "postgresql:///mydb";
  // Suffix the SVPOR signature with a per-run id — the
  // malu$fact_active_window_excl EXCLUDE constraint refuses two
  // active facts with overlapping valid windows on the same
  // (subject, verb).
  const run = randomBytes(4).toString("hex");
  const subject = `api_gateway_${run}`;

  const client = await MaluDBClient.connect({ connectionString: dsn });
  try {
    console.log(`connected to ${dsn} — maludb_core ${await client.version()}`);
    console.log(`run-id = ${run}, subject = ${subject}`);

    const sp = await client.registerSourcePackage({
      sourceType: "log",
      contentText: `node-example-01 [${run}]: 14:22Z api-gateway 5xx burst`,
      originJsonb: { uri: `log://oncall/node-example-01/${run}` },
    });
    console.log(`  source_package_id = ${sp}`);

    const c1 = await client.registerClaim({
      subject,
      verb: "observed",
      objectValue: "5xx_burst",
      statementText: `node-example-01 [${run}]: initial 5xx surge at 14:22Z`,
      sourcePackageId: sp,
    });
    const c2 = await client.registerClaim({
      subject,
      verb: "timed_out",
      objectValue: "health_probe",
      statementText: `node-example-01 [${run}]: health probe exceeded 2s`,
      sourcePackageId: sp,
    });
    console.log(`  claim_ids = ${c1}, ${c2}`);

    const f1 = await client.registerFact({
      claimIds: [c1, c2],
      subject,
      verb: "incident",
      objectValue: "latency_breach",
      statementText: `node-example-01 [${run}]: latency SLO breach`,
      verificationMethod: "oncall_review",
    });
    console.log(`  fact_id = ${f1}`);

    const ep = await client.registerEpisode({
      episodeKind: "incident",
      title: `node-example-01-outage-${run}`,
      summary: "Driver example outage",
      payload: { subject_class: subject, environment: "prod" },
    });
    console.log(`  episode_id = ${ep}`);

    console.log(`\n=== textSearch('${subject}') ===`);
    for (const h of await client.textSearch(subject, { limit: 5 })) {
      console.log(`  ${h.object_type.padEnd(14)} id=${String(h.object_id).padStart(5)}  rank=${h.rank.toFixed(4)}`);
    }

    console.log(`\n=== retrieve('${subject}') ===`);
    for (const h of await client.retrieve(subject, { limit: 5 })) {
      console.log(`  ${h.object_type.padEnd(14)} id=${String(h.object_id).padStart(5)}  strategy=${h.strategy}`);
    }

    console.log(`\n=== replayEpisode ===`);
    const envelope = await client.replayEpisode(ep, "current_valid");
    console.log(`  mode=${envelope.mode}  step_count=${envelope.steps.length}  evidence=${envelope.supporting_evidence.length}`);
  } finally {
    await client.end();
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
